JIT: add pass to merge common throw helper calls

AndyAyersMS opened the pull request:
Look for blocks with single statement noreturn calls, and try to reroute
flow so there's just one block call that all predecessors target.

Resolves #14770.

Note this impairs debuggability of optimized code a bit, as it can change which
line of code apparently invokes a throw helper in a backtrace. But since we're
already commoning jit-inserted throw helpers (like array index OOB) this is not
breaking any new ground.

We could also handle commoning BBJ_THROW blocks, with some extra effort,
but prototyping indicates duplicate throws are pretty rare.

AndyAyersMS committed:
JIT: add pass to merge common throw helper calls

Look for blocks with single statement noreturn calls, and try to reroute
flow so there's just one block call that all predecessors target.

Resolves #14770.

Note this impairs debuggability of optimized code a bit, as it can change which
line of code apparently invokes a throw helper in a backtrace. But since we're
already commoning jit-inserted throw helpers (like array index OOB) this is not
breaking any new ground.

We could also handle commoning BBJ_THROW blocks, with some extra effort,
but prototyping indicates duplicate throws are pretty rare.

AndyAyersMS commented:
@dotnet/jit-contrib PTAL

jit-diffs shows this hits modestly often...
```
Total bytes of diff: -43903 (-0.11% of base)
    diff is an improvement.

Top file improvements by size (bytes):
       -7399 : System.Private.CoreLib.dasm (-0.17% of base)
       -4746 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.08% of base)
       -4541 : System.Memory.dasm (-1.95% of base)
       -2464 : CommandLine.dasm (-0.55% of base)
       -2131 : Microsoft.CodeAnalysis.dasm (-0.12% of base)

82 total files with size differences (82 improved, 0 regressed), 47 unchanged.

Top method regressions by size (bytes):
          29 ( 1.67% of base) : Microsoft.CodeAnalysis.dasm - <DescendantTriviaIntoTrivia>d__161:MoveNext():bool:this
          17 ( 0.86% of base) : System.Security.Cryptography.Algorithms.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
          17 ( 0.86% of base) : System.Security.Cryptography.Cng.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
          13 ( 0.07% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:ReportOverloadResolutionFailureForASingleCandidate(ref,ref,int,byref,struct,struct,bool,bool,bool,bool,ref,ref,bool,ref,ref):this
           4 ( 1.04% of base) : System.Private.CoreLib.dasm - TextInfo:AddTitlecaseLetter(byref,byref,int,int):int:this

Top method improvements by size (bytes):
        -438 (-4.46% of base) : System.Memory.dasm - ReadOnlySequenceDebugView`1:.ctor(struct):this (7 methods)
        -396 (-5.97% of base) : Microsoft.CodeAnalysis.dasm - ArrayBuilderExtensions:SelectAsArray(ref,ref):struct (7 methods)
        -386 (-6.88% of base) : System.Memory.dasm - SequenceReader`1:TryCopyMultisegment(struct):bool:this (6 methods)
        -294 (-1.13% of base) : Microsoft.CodeAnalysis.dasm - AsyncQueue`1:WithCancellation(ref,struct):ref (49 methods)
        -294 (-1.80% of base) : System.Linq.Parallel.dasm - ParallelQuery`1:OfType():ref:this (49 methods)

Top method regressions by size (percentage):
          29 ( 1.67% of base) : Microsoft.CodeAnalysis.dasm - <DescendantTriviaIntoTrivia>d__161:MoveNext():bool:this
           4 ( 1.04% of base) : System.Private.CoreLib.dasm - TextInfo:AddTitlecaseLetter(byref,byref,int,int):int:this
          17 ( 0.86% of base) : System.Security.Cryptography.Algorithms.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
          17 ( 0.86% of base) : System.Security.Cryptography.Cng.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
           3 ( 0.33% of base) : xunit.core.dasm - MemberDataAttributeBase:GetData(ref):ref:this

Top method improvements by size (percentage):
         -56 (-17.50% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - LookupResult:CompactSymbols(int):this
        -278 (-16.27% of base) : System.Reflection.Metadata.dasm - MetadataBuilder:GetRowCounts():struct:this
         -56 (-16.18% of base) : Microsoft.CodeAnalysis.dasm - ExceptionHandlerContainerScope:CloseScope(ref):this
         -11 (-14.86% of base) : System.Diagnostics.Process.dasm - NtProcessManager:ReadCounterValue(int,struct):long
         -41 (-14.80% of base) : System.Private.CoreLib.dasm - Decimal:ToDecimal(struct):struct

1787 total methods with size differences (1781 improved, 6 regressed), 201157 unchanged.
```

erozenfeld reviewed (approved):
LGTM
What's the throughput impact of this optimization?

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -25694,3 +25694,211 @@ bool Compiler::fgUseThrowHelperBlocks()
+    // We will run two passes here, so we can avoid pred lists.
+    int numCandidates = 0;
+
+    // Scan for THROW blocks. Note early on throw helpers are not
```
"Note early on throw" more clear as "Note that early in compilation, throw" ?

AndyAyersMS replied:
Reworded.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -25694,3 +25694,211 @@ bool Compiler::fgUseThrowHelperBlocks()
+//    that can be merged, and opportunistically merges them.
+//
+//    Does not handle throws yet as the analysis is more costly and less
+//    likely to pay off.
```
I think it would be useful to give a code example here in the header comment of the pattern you are looking for, and describe why this is valuable.

AndyAyersMS replied:
Done.

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -16687,6 +16687,8 @@ void Compiler::fgMorph()
 
     fgUpdateFinallyTargetFlags();
 
+    fgTailMergeThrows();
```
Seems like you should make this a real PHASE so its cost gets counted separately.

AndyAyersMS replied:
Done.

BruceForstall reviewed (changes requested):
Why do you see regressions?

mikedn commented on src/jit/flowgraph.cpp:
```diff
@@ -25694,3 +25694,211 @@ bool Compiler::fgUseThrowHelperBlocks()
+    struct ThrowHelperKeyFuncs
+    {
+    public:
+        static bool Equals(const ThrowHelperKey& x, const ThrowHelperKey& y)
```
These 2 can be moved to `ThrowHelperKey` to avoid the extra noise from a separate class.

AndyAyersMS replied:
Merged the classes.

mikedn commented on src/jit/flowgraph.cpp:
```diff
@@ -25694,3 +25694,211 @@ bool Compiler::fgUseThrowHelperBlocks()
+
+    CompAllocator allocator(getAllocator(CMK_Generic));
+    typedef JitHashTable<ThrowHelperKey, ThrowHelperKeyFuncs, BasicBlock*> CallToBlockMap;
+    CallToBlockMap*  callMap  = new (allocator) CallToBlockMap(allocator);
```
These 2 do not need to be dynamically allocated.

AndyAyersMS replied:
Thanks, updated.

mikedn commented on src/jit/flowgraph.cpp:
```diff
@@ -25694,3 +25694,211 @@ bool Compiler::fgUseThrowHelperBlocks()
+        }
+    };
+
+    CompAllocator allocator(getAllocator(CMK_Generic));
```
Might want to add a new CMK value to get an idea how much memory this uses.

AndyAyersMS replied:
Should be pretty minimal, though of course that is not guaranteed. There aren't any examples in framework code with more than 20 or so throw helper calls.

AndyAyersMS replied:
Added ... have not looked at how big this can get, but still think it will be pretty; minimal,  In simple examples it's 200 bytes.

mikedn commented on src/jit/flowgraph.cpp:
```diff
@@ -25694,3 +25694,211 @@ bool Compiler::fgUseThrowHelperBlocks()
+    public:
+        static bool Equals(const ThrowHelperKey& x, const ThrowHelperKey& y)
+        {
+            return BasicBlock::sameEHRegion(x.m_block, y.m_block) && GenTreeCall::Equals(x.m_call, y.m_call);
```
Beware that the `GenTree` equality check ignores `gtCallMoreFlags`, most of `gtFlags` and has at least one bug (I ran into it a few days ago - it ignores the size of `GT_BLK` nodes). I'm going to fix the bug soon but it also turns out that `GenTree::Compare` is not widely used in the JIT so it's possible to run into more issues when a new use is added.

mikedn replied:
And if you were perhaps considering to use `gtHashValue` to deal with the more general case - that one also has some issues and it's only used by debug functionality.

AndyAyersMS replied:
Good point. Hadn't really looked closely at the compare code.

For the more general case where we try and merge throws (see linked issue) I wasn't relying on tree hashes.

AndyAyersMS commented:
The flow updates in this PR aren't sufficiently general yet. We need to ensure that the throw helper blocks can't be reached both by fall through and jump or else handle that case properly. For instance in `TextInfo:AddTitlecaseLetter(byref,byref,int,int):int:this` we see flow like:
```
------------ BB16 [005..006) -> BB18 (cond), preds={} succs={BB17,BB18}

***** BB16
STMT00041 (IL 0x005...  ???)
               [000175] ------------              *  JTRUE     void  
               [000174] ------------              \--*  NE        int   
               [000172] ------------                 +--*  LCL_VAR   int    V03 arg3         
               [000173] ------------                 \--*  CNS_INT   int    0

------------ BB17 [005..006), preds={} succs={BB18}

------------ BB18 [005..006), preds={} succs={BB19}

***** BB18
STMT00042 (IL 0x005...  ???)
               [000177] I-C-G-------              *  CALL      void   ThrowHelper.ThrowArgumentOutOfRangeException (exactContextHnd=0x00000000D1FFAB1E)
               [000176] ------------ arg0         \--*  CNS_INT   int    33
```
where `BB18` is non-canonical.

Need to think more about how to fix flow this early in the jit without requiring full pred list info  (computing even cheap preds is probably more costly than this entire phase -- I think we can instead leverage the jump target bb flag we set up early).

A couple other things to sort through:
* We really should clean out the trees in non-canonical jumped-to throw helper blocks, otherwise morph still processes the calls, even though the blocks are unreachable. 
* No block can be cleaned unless we're sure all predecessors have been rerouted, so we probably need to handle more predecessor types, just in case someone gets creative with switches and gotos. Will need to work up some test cases.
* Might as well flag when we introduce no return calls during importation, so we can avoid running this phase when there's nothing to find, which is most of the time (currently only ~1% or so of methods have these throw helper calls).

Back to the drawing board, for now.

AndyAyersMS committed:
Review feedback I

* Unify helper classes
* Stack allocate map headers
* Phasify
* Implement an earlyout scheme

AndyAyersMS committed:
Review Feedback II

Revise this phase to run just before `optOptimizeFlow`, so that we can leverage
the ref counts and predecessor lists to ensure we make correct flow updates.

Don't bother trying to clean out IR, let that happen naturally from as blocks
become unreferenced.

AndyAyersMS commented:
Moving it later in the phase list looks promising (now have it just before `optOptimizeFlow`). Flow updates now viable. Code size savings is now 48659 (up from 43903 above) as we can handle a wider variety of cases.

Speaking of variety, one odd case that's come up is that we may decide to tail call a throw helper. My merging code doesn't currently cope with this case. 

When that happens we don't mark the block as rare and it doesn't get moved out of line like a called throw helper does.  So we can end up with notreturn blocks in the middle of loops and such. We could probably fix that part easily enough.

But aside from that, it seems likes we should either never tail call throw helpers (if we tail call, we lose info about which method may have really caused the throw) or almost always tail call them (when we can). Never tail calling them allows even more merging but also a number of regressions in methods where merging isn't possible. Hmmm...

AndyAyersMS committed:
Heuristic for the tail call case

Suppress tail calling noreturn methods if there is more than one such
call site in the method, hoping that instead we can merge the calls.

AndyAyersMS commented:
Running via pin on crossgen of SPC, throughput impact is basically zero: 
```
baseline: 17,104,524,933
diff:     17,102,043,059
```

AndyAyersMS commented:
Going to reopen this -- main outstanding issue is how much we can trust tree comparison. The framework throw helper calls tend to have very simple argument trees, but that may not be the case elsewhere.

Added a simple heuristic to not tail call throw helpers if there's more than one such call in a method, and instead hope merging kicks in.

Still need to add some test cases for odd flow patterns.

Latest jit-diffs:
```
Total bytes of diff: -51919 (-0.13% of base)
    diff is an improvement.

Top file improvements by size (bytes):
       -7639 : System.Private.CoreLib.dasm (-0.17% of base)
       -4829 : System.Memory.dasm (-2.08% of base)
       -4749 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.08% of base)
       -4593 : System.Collections.Immutable.dasm (-0.42% of base)
       -2930 : System.Private.Xml.dasm (-0.08% of base)

83 total files with size differences (83 improved, 0 regressed), 46 unchanged.

Top method regressions by size (bytes):
          29 ( 1.67% of base) : Microsoft.CodeAnalysis.dasm - <DescendantTriviaIntoTrivia>d__161:MoveNext():bool:this
          17 ( 0.86% of base) : System.Security.Cryptography.Algorithms.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
          17 ( 0.86% of base) : System.Security.Cryptography.Cng.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
          13 ( 0.07% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:ReportOverloadResolutionFailureForASingleCandidate(ref,ref,int,byref,struct,struct,bool,bool,bool,bool,ref,ref,bool,ref,ref):this
           4 ( 3.48% of base) : Microsoft.CodeAnalysis.dasm - ObjectReader:CreateInstance(ref):ref:this

Top method improvements by size (bytes):
        -588 (-4.60% of base) : System.Collections.Immutable.dasm - Node:CopyTo(ref,int):this (28 methods)
        -438 (-4.46% of base) : System.Memory.dasm - ReadOnlySequenceDebugView`1:.ctor(struct):this (7 methods)
        -396 (-5.97% of base) : Microsoft.CodeAnalysis.dasm - ArrayBuilderExtensions:SelectAsArray(ref,ref):struct (7 methods)
        -386 (-6.88% of base) : System.Memory.dasm - SequenceReader`1:TryCopyMultisegment(struct):bool:this (6 methods)
        -294 (-1.13% of base) : Microsoft.CodeAnalysis.dasm - AsyncQueue`1:WithCancellation(ref,struct):ref (49 methods)

Top method regressions by size (percentage):
           4 ( 3.48% of base) : Microsoft.CodeAnalysis.dasm - ObjectReader:CreateInstance(ref):ref:this
          29 ( 1.67% of base) : Microsoft.CodeAnalysis.dasm - <DescendantTriviaIntoTrivia>d__161:MoveNext():bool:this
          17 ( 0.86% of base) : System.Security.Cryptography.Algorithms.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
          17 ( 0.86% of base) : System.Security.Cryptography.Cng.dasm - PasswordBasedEncryption:Encrypt(struct,struct,ref,bool,struct,ref,struct,ref,struct):int
           3 ( 0.33% of base) : xunit.core.dasm - MemberDataAttributeBase:GetData(ref):ref:this

Top method improvements by size (percentage):
         -56 (-17.50% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - LookupResult:CompactSymbols(int):this
        -147 (-16.67% of base) : System.Collections.Immutable.dasm - ImmutableList`1:Sort(int,int,ref):ref:this (7 methods)
        -278 (-16.27% of base) : System.Reflection.Metadata.dasm - MetadataBuilder:GetRowCounts():struct:this
         -19 (-16.24% of base) : System.Net.Http.dasm - WinHttpHandler:SetSessionHandleConnectionOptions(ref):this
         -56 (-16.18% of base) : Microsoft.CodeAnalysis.dasm - ExceptionHandlerContainerScope:CloseScope(ref):this

2643 total methods with size differences (2637 improved, 6 regressed), 200301 unchanged.
```

AndyAyersMS committed:
trivial edit to pick up latest changes

AndyAyersMS commented:
Going to bounce this to trigger retest.

AndyAyersMS commented:
Ah, looks like I need to rebase...

AndyAyersMS committed:
update for recent changes in stmt

CarolEidt reviewed (commented):
I have mostly questions.

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,340 @@ bool Compiler::fgUseThrowHelperBlocks()
+//
+//    Also note as a quirk of how we model flow that both BB17 and BB21
+//    have successor blocks. We don't turn these into true throw blocks
+//    until morph.
```
Yes.

Note we only ever see this case when optimizing (since recognition of noreturn calls currently requires a failed inline attempt) and any blocks made unreachable during morph get cleaned up pretty quickly.

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,340 @@ bool Compiler::fgUseThrowHelperBlocks()
+//
+// Notes:
+//    Alters fall through flow of predBlock so it jumps to the canonicalBlock
+//    via a new basic block.
```
Ideally there'd be a utility to call that would check for special restrictions, reverse the branch condition and make suitable flow updates. The logic for this is is in `fgUpdateFlowGraph` and nontrivial so I didn't want to duplicate code, and the blocks get updated (if suitable) right after this phase, during `optOptimizeFlow`.

I could refactor, if you think it makes sense to try and optimize this on the spot instead of waiting...

CarolEidt replied:
No, it sounds like this is the right thing to do, but A comment to that effect would help explain why it's being done this way.

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,340 @@ bool Compiler::fgUseThrowHelperBlocks()
+        if (callMap.Lookup(key, &canonicalBlock))
+        {
+            // Make sure we're not at risk of crossing EH boundaries
+            assert(BasicBlock::sameEHRegion(block, canonicalBlock));
```
Omitted.

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,340 @@ bool Compiler::fgUseThrowHelperBlocks()
+    for (BasicBlock* block = fgLastBB; block != nullptr; block = block->bbPrev)
+    {
+        // Hmm, odd BBJ_RETURN case when throw helper is tail called.
+        // TODO: sort this out over in morph.
```
It's more of a workaround, but yes there is some code over in morph now.

AndyAyersMS commented on src/jit/compiler.h:
```diff
@@ -6367,6 +6377,18 @@ class Compiler
+        optNoReturnCallCount++;
+    }
+
+    unsigned optNoReturnCallCount;
```
Fixed

AndyAyersMS committed:
Review Feedback III

CarolEidt reviewed (approved):
LGTM - thanks for the explanations

AndyAyersMS commented:
Looking at `GenTreeCall::Compare`, there are quite a few fields that don't factor in:
```
fgArgInfo
regArgListCount
regArgList
callSig
gtReturnTypeDesc
gtOtherRegs
gtSpillFlags
gtFlags
gtCallMoreFlags
gtReturnType
... the big union ...
gtEntryPoint (partial)
```
Not sure what to do about this yet. Some of the information is partially redundant so current comparisons may be sufficient (eg return type will be determined by the call target).

briansull commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,330 @@ bool Compiler::fgUseThrowHelperBlocks()
+    // First pass
+    //
+    // Scan for THROW blocks. Note early on in compilation (before morph)
+    // noretrurn blcoks are not marked as BBJ_THROW.
```
spelling: blcoks

briansull commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,330 @@ bool Compiler::fgUseThrowHelperBlocks()
+
+        // Have we found an equivalent call already?
+        ThrowHelper key(block, call);
+        if (callMap.Lookup(key, &canonicalBlock))
```
Do you need a check that insures that both blocks are in the same try region?

AndyAyersMS replied:
The check is there, it's up in the `ThrowHelpers::Equals` method.

AndyAyersMS committed:
Review Feedback IV plus test case

AndyAyersMS commented:
Diff summary for the test case methods. Note not using tail calls for throw helpers hurts code size slightly, in some cases.
```
PMI Diffs for C:\repos\coreclr3\tests\src\JIT\opt\ThrowHelper\ThrowHelper.exe for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -65 (-1.82% of base)
    diff is an improvement.
Top file improvements by size (bytes):
         -65 : ThrowHelper.dasm (-1.82% of base)
1 total files with size differences (1 improved, 0 regressed), 0 unchanged.
Top method regressions by size (bytes):
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_IfOneTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_IfTwoTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_GotoTwoTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_SwitchOneTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_SwitchTwoTail(int)
Top method improvements by size (bytes):
         -18 (-38.30% of base) : ThrowHelper.dasm - TestCases:ThreeIdenticalThrowHelpers_IfOneTail(int)
         -15 (-26.79% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpersSameArgTrees(int,ref):int
         -12 (-27.91% of base) : ThrowHelper.dasm - TestCases:ThreeIdenticalThrowHelpers_If(int):int
          -9 (-27.27% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_GotoOneTail(int)
          -9 (-27.27% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_SwitchGotoOneTail(int)
Top method regressions by size (percentage):
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_IfOneTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_IfTwoTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_GotoTwoTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_SwitchOneTail(int)
           4 (20.00% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_SwitchTwoTail(int)
Top method improvements by size (percentage):
         -18 (-38.30% of base) : ThrowHelper.dasm - TestCases:ThreeIdenticalThrowHelpers_IfOneTail(int)
         -12 (-27.91% of base) : ThrowHelper.dasm - TestCases:ThreeIdenticalThrowHelpers_If(int):int
          -9 (-27.27% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_GotoOneTail(int)
          -9 (-27.27% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpers_SwitchGotoOneTail(int)
         -15 (-26.79% of base) : ThrowHelper.dasm - TestCases:TwoIdenticalThrowHelpersSameArgTrees(int,ref):int
17 total methods with size differences (11 improved, 6 regressed), 21 unchanged.
```

AndyAyersMS commented:
@BruceForstall want to take another look?

There is still some concern about the potential for bugs in tree comparison (mainly false matches from insufficiently detailed compares). If we think this is a deal breaker I'll probably shelve this and try and address that separately.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,333 @@ bool Compiler::fgUseThrowHelperBlocks()
+    if (updateCount > 0)
+    {
+        assert(fgModified);
+        fgModified = false;
```
This seems a little worrisome, but ok... Wouldn't (Shouldn't) the first code to generate "flow-dependent side data" reset this?

AndyAyersMS replied:
Similar to logic in `fgCreateFuncletPrologBlocks`. 

The `fgModified` concept should probably be updated to just directly invalidate the dependent analyses, if any (in this case and the funclet case, there are none).

BruceForstall reviewed (approved):
A few nits, but LGTM

AndyAyersMS committed:
review feedback V

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -25718,3 +25720,333 @@ bool Compiler::fgUseThrowHelperBlocks()
+    // We will run two passes here.
+    //
+    // The first pass finds candidate blocks and builds the mapping
+    // from candidates to canonical blocks, the second pass modifies flow.
```
Revised this to describe what canonical means here.

AndyAyersMS commented:
See also dotnet/runtime#35135.
